### ISSUE TITLE
Allow multiple game instances for testing

### ISSUE THREAD

**Issue by MulleDK19** [Facepunch/garrysmod-requests#991]
Working on a gamemode and (unrelated) addon, and auto refresh is extremely useful.
However, it's cumbersome to test multiplayer related code because you need a real person to help you. And asking friends to test, is a good way to lose friends.

And since bots can't do everything, they're not useful for most scenarios.

It'd be nice if we could run multiple game instances on the same PC for testing purposes. I understand there'd be a couple of complications like `player:SteamID()`, `player:SteamID64()`, etc. but these could be changed to append a value, eg. `STEAM_0:0:1234567_2`, `STEAM_0:0:1234567_3`, etc.

This could be accompanied by something like `sv_allowduplicateplayers` or `sv_allow_multiple_connections_from_same_user_without_forcing_them_to_lose_friends`

This wouldn't just be a little helpful, it'd be extremely helpful.

As a use case, I'm currently working on a custom prop protection, and currently working on allowing players to request permissions from other players, which other players can then grant or deny. And this is a nightmare because I need to scrape the last few friends to test.

**Comment by Kefta** [comment]
You can use the -multirun startup command to do this.

**Comment by MulleDK19** [comment]
@Kefta Cool, thanks, didn't know that.

Although, that doesn't fix the issue of identical SteamIDs.
Although, I guess I could hook player:GetSteamID(), etc. to return different IDs myself.

**Comment by robotboy655** [comment]
This is already possible, sv_lan 1 to allow "duplicate steam ids".

**Comment by MulleDK19** [comment]
@robotboy655 It has several issues, so this request is still relevant. Eg. calling Kick() on a duplicate will kick the first instance of the player who joined.

For now, I've written this hook to provide different SteamIDs to duplicate players:
```lua
//////////
///  FOR DEBUGGING
local PlayerMetatable = FindMetaTable("Player")
OldPlayerMetatableSteamID = OldPlayerMetatableSteamID or PlayerMetatable.SteamID
OldPlayerMetatableSteamID64 = OldPlayerMetatableSteamID64 or PlayerMetatable.SteamID64

function PlayerMetatable.SteamID(self)
	-- FIXME: When a player with duplicate (either the player themself or a duplicate) leaves, players get new SteamIDs.
	
	-- On clients, duplicate players have the same SteamID.
	-- Unfortunately, on the server, the duplicates have STEAM_0:0:0.
	-- We'll make up a fake SteamID based on the player's index so they're the same on both server and clients.
	-- First instance of the player to join gets their real SteamID, the rest gets the real SteamID appended by '_X' where X is the player index.
	
	if !SERVER then
		-- NOTE: On clients, each player duplicate has the real player's real SteamID.
		local actualSteamID = OldPlayerMetatableSteamID(ply)
		local actualPlayer = nil
		for i, k in pairs(player.GetAll()) do
			if OldPlayerMetatableSteamID(k) == actualSteamID then
				actualPlayer = k
				break
			end
		end
		
		if ply == actualPlayer then
			-- This is the actual player, return the actual SteamID.
			return actualSteamID
		else
			-- This is a duplicate of the player, return a fake SteamID.
			return actualSteamID .. "_" .. tostring(ply:EntIndex())
		end
	else
		-- NOTE: However, on the server, only the first instance of the player has the player's real SteamID. The duplicates have STEAM_0:0:0.
		-- Since there can be duplicates of different players we need to find out which player the duplicate belongs to.
		if OldPlayerMetatableSteamID(ply) == "STEAM_0:0:0" then
			-- This is a duplicate. Find out who this is a duplicate player of.
			local ipAddressAndPort = ply:IPAddress()
			local ipStart, ipEnd = string.find(ipAddressAndPort, ":")
			local actualIpAddress = string.sub(ipAddressAndPort, 1, ipStart - 1)
			local actualPlayer = nil
			local actualSteamID = nil
			for i, k in pairs(player.GetAll()) do
				ipAddressAndPort = k:IPAddress()
				ipStart, ipEnd = string.find(ipAddressAndPort, ":")
				local ipAddress = string.sub(ipAddressAndPort, 1, ipStart - 1)
				if ipAddress == actualIpAddress then
					actualPlayer = k
					actualSteamID = OldPlayerMetatableSteamID(k)
					break
				end
			end
			
			-- This is a duplicate of the player, return a fake SteamID.
			return actualSteamID .. "_" .. ply:EntIndex()
		else
			-- This is an actual player.
			return OldPlayerMetatableSteamID(ply)
		end
	end
end

function PlayerMetatable.SteamID64(self)
	-- Since SteamID64() returns a string, and that number can't accurately be represented by a double anyway, just return SteamID().
	return self:SteamID()
end
///  FOR DEBUGGING
//////////```

**Comment by robotboy655** [comment]
Cannot have more than 1 of the same SteamID on the server. That is not going to change. And we cannot give duplicate clients someone else's SteamIDs for obvious reasons.

**Comment by neico** [comment]
There are multiple ways to store data for players with the same SteamID.

You could try UniqueID which *should* give you a different result even on multiple instances, then there's also a per-IP approach (which in your case might not actually work, you'd need more of a client port approach, not sure if you can get that without a c++ module)

If you really want to go the SteamID way, make full use of the SteamID space.  
Gmod uses steam universe 0 today, newer source1 games use 1, you could choose to use your own universe id, like 10. (there can be up to 256 universes)  
There's also account types, AnonUser or MultiSeat might be something of interest, or yet again, make your own type. (there can be up to 16 types)  
See https://developer.valvesoftware.com/wiki/SteamID

So in summary, there's lots of room to give multiple instances really unique id's. (if those kind of SteamID's are supported by all addons is another topic tho as some will probably expect at least `STEAM_0:` which is quite stupid imo)
